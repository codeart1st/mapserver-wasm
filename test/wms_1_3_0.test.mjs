import { loadMapServer, initMapServer, getDataFromResByteBuffer, printMapServerLog } from './utils'

let mapserver

beforeAll(async () => {
  mapserver = await loadMapServer()

  initMapServer(mapserver, {
    mapFileName: 'wms.map'
  })
})

afterAll(() => {
  //printMapServerLog(mapserver, 'error.log')
})

describe('WMS 1.3.0', () => {
  const globalParams = 'SERVICE=WMS&VERSION=1.3.0'

  test('GetCapabilities request should be valid', () => {
    const { cwrap } = mapserver

    const dispatchRequest = cwrap('dispatchRequest', 'string', ['string', 'string'])
    const searchParams = new URLSearchParams(globalParams)

    searchParams.set('REQUEST', 'GetCapabilities')

    const response = dispatchRequest('GET', searchParams.toString())

    expect(response).toMatchSnapshot()
  })

  test('GetMap request should be return a jpeg image', () => {
    const { cwrap } = mapserver

    const dispatchRequestBytes = cwrap('dispatchRequestBytes', 'number', ['string', 'string'])
    const searchParams = new URLSearchParams(globalParams)

    searchParams.set('REQUEST', 'GetMap')
    searchParams.set('WIDTH', '1000')
    searchParams.set('HEIGHT', '1000')
    searchParams.set('LAYERS', 'testraster')
    searchParams.set('CRS', 'EPSG:3857')
    searchParams.set('BBOX', '-20037508,-20037508,20037508,20037508')
    searchParams.set('FORMAT', 'image/jpeg')

    const resByteBuffer = dispatchRequestBytes('GET', searchParams.toString())
    const body = getDataFromResByteBuffer(mapserver, resByteBuffer)

    expect(body).toMatchFileSnapshot()
  })

  test('GetMap request should be return a png image', () => {
    const { cwrap } = mapserver

    const dispatchRequestBytes = cwrap('dispatchRequestBytes', 'number', ['string', 'string'])
    const searchParams = new URLSearchParams(globalParams)

    searchParams.set('REQUEST', 'GetMap')
    searchParams.set('WIDTH', '1000')
    searchParams.set('HEIGHT', '1000')
    searchParams.set('LAYERS', 'testraster')
    searchParams.set('CRS', 'EPSG:3857')
    searchParams.set('BBOX', '-20037508,-20037508,20037508,20037508')
    searchParams.set('FORMAT', 'image/png')

    const resByteBuffer = dispatchRequestBytes('GET', searchParams.toString())
    const body = getDataFromResByteBuffer(mapserver, resByteBuffer)

    expect(body).toMatchFileSnapshot()
  })

  test('GetMap request should transform to different srs', () => {
    const { cwrap } = mapserver

    const dispatchRequestBytes = cwrap('dispatchRequestBytes', 'number', ['string', 'string'])
    const searchParams = new URLSearchParams(globalParams)

    searchParams.set('REQUEST', 'GetMap')
    searchParams.set('WIDTH', '1000')
    searchParams.set('HEIGHT', '1000')
    searchParams.set('LAYERS', 'testraster')
    searchParams.set('CRS', 'EPSG:4326')
    searchParams.set('BBOX', '-180.0000,-90.0000,180.0000,90.0000')
    searchParams.set('FORMAT', 'image/jpeg')

    const resByteBuffer = dispatchRequestBytes('GET', searchParams.toString())
    const body = getDataFromResByteBuffer(mapserver, resByteBuffer)

    expect(body).toMatchFileSnapshot()
  })

  test('GetMap request should print exception to image output', () => {
    const { cwrap } = mapserver

    const dispatchRequestBytes = cwrap('dispatchRequestBytes', 'number', ['string', 'string'])
    const searchParams = new URLSearchParams(globalParams)

    searchParams.set('REQUEST', 'GetMap')
    searchParams.set('WIDTH', '256')
    searchParams.set('HEIGHT', '256')
    searchParams.set('LAYERS', 'foo')
    searchParams.set('CRS', 'EPSG:4326')
    searchParams.set('BBOX', '-180.0000,-90.0000,180.0000,90.0000')
    searchParams.set('FORMAT', 'image/png')
    searchParams.set('EXCEPTIONS', 'INIMAGE')

    const resByteBuffer = dispatchRequestBytes('GET', searchParams.toString())
    const body = getDataFromResByteBuffer(mapserver, resByteBuffer)

    expect(body).toMatchFileSnapshot()
  })
})