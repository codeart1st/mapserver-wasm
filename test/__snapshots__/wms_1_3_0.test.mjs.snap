// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WMS 1.3.0 > GetCapabilities request should be valid 1`] = `
"Content-Type: text/xml; charset=UTF-8

<?xml version='1.0' encoding="UTF-8" standalone="no" ?>
<WMS_Capabilities version="1.3.0"  xmlns="http://www.opengis.net/wms"   xmlns:sld="http://www.opengis.net/sld"   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"   xmlns:ms="http://mapserver.gis.umn.edu/mapserver"   xsi:schemaLocation="http://www.opengis.net/wms http://schemas.opengis.net/wms/1.3.0/capabilities_1_3_0.xsd  http://www.opengis.net/sld http://schemas.opengis.net/sld/1.1.0/sld_capabilities.xsd  http://mapserver.gis.umn.edu/mapserver http://localhost?service=WMS&amp;version=1.3.0&amp;request=GetSchemaExtension">
<Service>
  <Name>WMS</Name>
<!-- WARNING: Mandatory metadata "wms_title" or "ows_title" was missing in this context. -->
  <Title>MS</Title>
  <OnlineResource xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://localhost?"/>
  <ContactInformation>
  </ContactInformation>
  <MaxWidth>4096</MaxWidth>
  <MaxHeight>4096</MaxHeight>
</Service>

<Capability>
  <Request>
    <GetCapabilities>
      <Format>text/xml</Format>
      <DCPType>
        <HTTP>
          <Get><OnlineResource xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://localhost?"/></Get>
          <Post><OnlineResource xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://localhost?"/></Post>
        </HTTP>
      </DCPType>
    </GetCapabilities>
    <GetMap>
      <Format>image/png</Format>
      <Format>image/jpeg</Format>
      <Format>image/png; mode=8bit</Format>
      <Format>image/vnd.jpeg-png</Format>
      <Format>image/vnd.jpeg-png8</Format>
      <Format>application/json</Format>
      <DCPType>
        <HTTP>
          <Get><OnlineResource xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://localhost?"/></Get>
          <Post><OnlineResource xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://localhost?"/></Post>
        </HTTP>
      </DCPType>
    </GetMap>
    <GetFeatureInfo>
      <Format>text/plain</Format>
      <Format>application/vnd.ogc.gml</Format>
      <DCPType>
        <HTTP>
          <Get><OnlineResource xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://localhost?"/></Get>
          <Post><OnlineResource xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://localhost?"/></Post>
        </HTTP>
      </DCPType>
    </GetFeatureInfo>
    <sld:DescribeLayer>
      <Format>text/xml</Format>
      <DCPType>
        <HTTP>
          <Get><OnlineResource xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://localhost?"/></Get>
          <Post><OnlineResource xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://localhost?"/></Post>
        </HTTP>
      </DCPType>
    </sld:DescribeLayer>
    <sld:GetLegendGraphic>
      <Format>image/png</Format>
      <Format>image/jpeg</Format>
      <Format>image/png; mode=8bit</Format>
      <Format>image/vnd.jpeg-png</Format>
      <Format>image/vnd.jpeg-png8</Format>
      <DCPType>
        <HTTP>
          <Get><OnlineResource xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://localhost?"/></Get>
          <Post><OnlineResource xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://localhost?"/></Post>
        </HTTP>
      </DCPType>
    </sld:GetLegendGraphic>
    <ms:GetStyles>
      <Format>text/xml</Format>
      <DCPType>
        <HTTP>
          <Get><OnlineResource xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://localhost?"/></Get>
          <Post><OnlineResource xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://localhost?"/></Post>
        </HTTP>
      </DCPType>
    </ms:GetStyles>
  </Request>
  <Exception>
    <Format>XML</Format>
    <Format>INIMAGE</Format>
    <Format>BLANK</Format>
  </Exception>
  <sld:UserDefinedSymbolization SupportSLD="1" UserLayer="0" UserStyle="1" RemoteWFS="0" InlineFeature="0" RemoteWCS="0"/>
  <Layer>
    <Name>MS</Name>
<!-- WARNING: Mandatory metadata "wms_title" or "ows_title" was missing in this context. -->
    <Title>MS</Title>
    <Abstract>MS</Abstract>
    <CRS>EPSG:3857</CRS>
    <CRS>EPSG:4326</CRS>
    <EX_GeographicBoundingBox>
        <westBoundLongitude>-0.000009</westBoundLongitude>
        <eastBoundLongitude>-0.000009</eastBoundLongitude>
        <southBoundLatitude>-0.000009</southBoundLatitude>
        <northBoundLatitude>-0.000009</northBoundLatitude>
    </EX_GeographicBoundingBox>
    <BoundingBox CRS="EPSG:3857"
                minx="-1" miny="-1" maxx="-1" maxy="-1" />
    <Layer queryable="0" opaque="0" cascaded="0">
        <Name>testraster</Name>
        <Title>testraster</Title>
        <CRS>EPSG:3857</CRS>
        <CRS>EPSG:4326</CRS>
        <EX_GeographicBoundingBox>
            <westBoundLongitude>-180.000000</westBoundLongitude>
            <eastBoundLongitude>180.000000</eastBoundLongitude>
            <southBoundLatitude>-85.051129</southBoundLatitude>
            <northBoundLatitude>85.051129</northBoundLatitude>
        </EX_GeographicBoundingBox>
        <BoundingBox CRS="EPSG:3857"
                    minx="-2.00375e+07" miny="-2.00375e+07" maxx="2.00375e+07" maxy="2.00375e+07" />
        <MetadataURL type="TC211">
          <Format>text/xml</Format>
          <OnlineResource xmlns:xlink="http://www.w3.org/1999/xlink" xlink:type="simple" xlink:href="http://localhost?request=GetMetadata&amp;layer=testraster"/>
        </MetadataURL>
    </Layer>
  </Layer>
</Capability>
</WMS_Capabilities>
"
`;
